pkg/console/util.go

```diff
@@ -325,25 +334,44 @@ func execute(ctx context.Context, g *gocui.Gui, env []string, cmdName string) er
 	if err != nil {
 		return err
 	}
+	defer stderr.Close()
+
 	stdout, err := cmd.StdoutPipe()
 	if err != nil {
 		return err
 	}
+	defer stdout.Close()
+
+	var wg sync.WaitGroup
+	var writeLock sync.Mutex
+
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		printToPanelAndLog(g, installPanel, "[stderr]", stderr, &writeLock)
+	}()
+
+	go func() {
+		defer wg.Done()
+		printToPanelAndLog(g, installPanel, "[stdout]", stdout, &writeLock)
+	}()
+
 	if err := cmd.Start(); err != nil {
 		return err
 	}
 
-	scanner := bufio.NewScanner(stdout)
-	scanner.Split(bufio.ScanLines)
-	for scanner.Scan() {
-		printToPanel(g, scanner.Text(), installPanel)
-	}
-	scanner = bufio.NewScanner(stderr)
-	scanner.Split(bufio.ScanLines)
+	wg.Wait()
+	return cmd.Wait()
+}
+
+func printToPanelAndLog(g *gocui.Gui, panel string, logPrefix string, reader io.Reader, lock *sync.Mutex) {
+	scanner := bufio.NewScanner(reader)
 	for scanner.Scan() {
-		printToPanel(g, scanner.Text(), installPanel)
+		logrus.Infof("%s: %s", logPrefix, scanner.Text())
+		lock.Lock()
```

Comment:
Just wondering if adding the mutex is necessary here. `printToPanel` uses `gocui.Gui.Update` to write data to panel so this event-based system should prevent the race-condition problem naturally.

Reply:
Probably not needed. It should be no harm to keep them.
BTW, the codes are "inspired" from https://github.com/rancher/system-agent/blob/b1f6aced9e8d3f491e4fea375265207a0eda6c61/pkg/applyinator/applyinator.go#L373-L470